contracts/contracts/multisig/Multisig.cairo

```diff
@@ -474,24 +436,113 @@ func execute_transaction{
     return (response_len=response.retdata_size, response=response.retdata)
 end
 
+# Sets number of required confirmations. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_confirmations_required.
+@external
+func set_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    confirmations_required : felt
+):
+    const lower_bound = 1
+
+    let (caller) = get_caller_address()
+    let (contract_address) = get_contract_address()
+
+    with_attr error_message("caller shall be multisig"):
+        assert caller = contract_address
+    end
+
+    let (owners_len) = _owners_len.read()
+
+    with_attr error_message("invalid number of required confirmations"):
+        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
+    end
+
+    _confirmations_required.write(confirmations_required)
+    ConfirmationsSet.emit(confirmations_required)
+    return ()
+end
+
+# Sets the owners field on the multisig. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_owners.
+# Number of required confirmations is decreased in case its larger than
+# number of new owners.
+@external
+func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    alloc_locals
+
+    let (caller) = get_caller_address()
+    let (contract_address) = get_contract_address()
+
+    with_attr error_message("caller shall be multisig"):
+        assert caller = contract_address
+    end
+
+    with_attr error_message("invalid num of owners"):
+        assert_not_zero(owners_len)
+    end
+
+    with_attr error_message("owners has to be unique"):
+        assert_unique_addresses(owners_len, owners)
+    end
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    let (lt) = is_le(owners_len, confirmations_required - 1)
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
```

Comment:
Event not emitted here, but they are when the `_confirmations_required` are changed directly. Consider using an internal fn for this logic and reuse so there are no gaps like this.

Reply:
Done